File: src/eval/evalSeeder.ts
/**
 * Eval fixture seeder — seeds and cleans up test data from eval/dataset.json fixtures.
 * Bypasses the LLM deduplication router and uses explicit IDs for deterministic eval runs.
 */

import { MeilisearchDB } from "../storage/meilisearchDB";
import { Embedder } from "../storage/embedder";
import type { MemoryCategory, MemoryOwner } from "../core/types";

export interface MemoryFixture {
  id: string;
  content: string;
  category: string;
  owner: string;
  importance: number;
  project: string;
  created_at?: string;
}

export interface SkillFixture {
  id: string;
  name: string;
  description: string;
  project: string;
  created_at?: string;
}

export interface ContextFixture {
  uri: string;
  name: string;
  abstract: string;
  overview?: string;
  content?: string;
  parent_uri?: string | null;
  project: string;
}

export interface FixtureSpec {
  memories?: MemoryFixture[];
  skills?: SkillFixture[];
  context?: ContextFixture[];
}

/**
 * Seeds all fixtures into Elasticsearch using explicit IDs.
 * Context nodes are inserted in topological order (parents before children).
 */
export async function seedFixtures(db: MeilisearchDB, fixtures: FixtureSpec, log?: (msg: string) => void): Promise<void> {
  const emit = log ?? (() => {});

  const memories = fixtures.memories ?? [];
  const skills = fixtures.skills ?? [];
  const contextNodes = topologicalSort(fixtures.context ?? []);

  emit(`Seeding ${memories.length} memories...`);
  for (const m of memories) {
    const embedding = await Embedder.getEmbedding(m.content);
    const ts = m.created_at ?? new Date().toISOString();
    await db.addMemory(
      {
        id: m.id,
        content: m.content,
        category: m.category as MemoryCategory,
        owner: m.owner as MemoryOwner,
        importance: m.importance,
        project: m.project,
        metadata: {},
        created_at: ts,
        updated_at: ts,
      },
      embedding
    );
    emit(`  ✓ memory ${m.id}`);
  }

  emit(`Seeding ${skills.length} skills...`);
  for (const s of skills) {
    const embedding = await Embedder.getEmbedding(`${s.name}: ${s.description}`);
    const ts = s.created_at ?? new Date().toISOString();
    await db.addSkill(
      {
        id: s.id,
        name: s.name,
        description: s.description,
        project: s.project,
        metadata: {},
        created_at: ts,
        updated_at: ts,
      },
      embedding
    );
    emit(`  ✓ skill ${s.id}`);
  }

  emit(`Seeding ${contextNodes.length} context nodes...`);
  for (const c of contextNodes) {
    const embedding = await Embedder.getEmbedding(`${c.name}: ${c.abstract}`);
    const ts = new Date().toISOString();
    await db.addContextNode(
      {
        uri: c.uri,
        name: c.name,
        abstract: c.abstract,
        overview: c.overview,
        content: c.content,
        parent_uri: c.parent_uri ?? null,
        project: c.project,
        metadata: {},
        created_at: ts,
        updated_at: ts,
      },
      embedding
    );
    emit(`  ✓ context ${c.uri}`);
  }

  emit("Seeding complete.");
}

/**
 * Deletes all seeded fixtures from Elasticsearch.
 * Context nodes are deleted in reverse topological order (leaves before parents).
 */
export async function cleanupFixtures(db: MeilisearchDB, fixtures: FixtureSpec, log?: (msg: string) => void): Promise<void> {
  const emit = log ?? (() => {});

  for (const m of fixtures.memories ?? []) {
    await db.deleteMemory(m.id);
    emit(`  ✗ deleted memory ${m.id}`);
  }

  for (const s of fixtures.skills ?? []) {
    await db.deleteSkill(s.id);
    emit(`  ✗ deleted skill ${s.id}`);
  }

  const reversed = topologicalSort(fixtures.context ?? []).reverse();
  for (const c of reversed) {
    await db.deleteContextNode(c.uri);
    emit(`  ✗ deleted context ${c.uri}`);
  }

  emit("Cleanup complete.");
}

/** Sort context nodes by URI depth so parents are inserted before children. */
function topologicalSort(nodes: ContextFixture[]): ContextFixture[] {
  return [...nodes].sort((a, b) => {
    const depthA = (a.uri.match(/\//g) ?? []).length;
    const depthB = (b.uri.match(/\//g) ?? []).length;
    return depthA - depthB;
  });
}

File: mairu/internal/eval/eval_seeder.go
package eval

import (
	"context"

	"mairu/internal/contextsrv"
)

type MemoryFixture struct {
	ID         string `json:"id"`
	Content    string `json:"content"`
	Category   string `json:"category"`
	Owner      string `json:"owner"`
	Importance int    `json:"importance"`
	Project    string `json:"project"`
}

type SkillFixture struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Project     string `json:"project"`
}

type ContextFixture struct {
	URI       string  `json:"uri"`
	Name      string  `json:"name"`
	Abstract  string  `json:"abstract"`
	Overview  string  `json:"overview"`
	Content   string  `json:"content"`
	ParentURI *string `json:"parent_uri"`
	Project   string  `json:"project"`
}

type FixtureSpec struct {
	Memories []MemoryFixture  `json:"memories"`
	Skills   []SkillFixture   `json:"skills"`
	Context  []ContextFixture `json:"context"`
}

func SeedFixtures(ctx context.Context, svc *contextsrv.AppService, spec FixtureSpec) error {
	for _, m := range spec.Memories {
		_, err := svc.CreateMemory(contextsrv.MemoryCreateInput{
			Project:    m.Project,
			Content:    m.Content,
			Category:   m.Category,
			Owner:      m.Owner,
			Importance: m.Importance,
		})
		if err != nil {
			return err
		}
	}
	for _, s := range spec.Skills {
		_, err := svc.CreateSkill(contextsrv.SkillCreateInput{
			Project:     s.Project,
			Name:        s.Name,
			Description: s.Description,
		})
		if err != nil {
			return err
		}
	}
	for _, c := range spec.Context {
		_, err := svc.CreateContextNode(contextsrv.ContextCreateInput{
			Project:   c.Project,
			URI:       c.URI,
			ParentURI: c.ParentURI,
			Name:      c.Name,
			Abstract:  c.Abstract,
			Overview:  c.Overview,
			Content:   c.Content,
		})
		if err != nil {
			return err
		}
	}
	return nil
}

func CleanupFixtures(ctx context.Context, svc *contextsrv.AppService, spec FixtureSpec) error {
	// Revers order for context nodes (children first) is ideal, but here we just delete by ID/URI
	for i := len(spec.Context) - 1; i >= 0; i-- {
		_ = svc.DeleteContextNode(spec.Context[i].URI)
	}
	for _, s := range spec.Skills {
		_ = svc.DeleteSkill(s.ID)
	}
	for _, m := range spec.Memories {
		_ = svc.DeleteMemory(m.ID)
	}
	return nil
}
